File: cli/backend/nws/nws_current.py
"""Alternative Weather Backend"""
import time
from datetime import datetime

from core.backend import WeatherCondition
from core.backend import WeatherData
from core.backend import WindData


class NationalWeatherServiceCurrent(WeatherData):
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, data, metric):
        self.metric = metric
        wind = WindData(
            self.convert_speed(data["properties"]["windSpeed"]["values"][0]["value"]),
            data["properties"]["windDirection"]["values"][0]["value"],
        )
        # print(data["properties"]["updateTime"])
        super().__init__()
        self.time = int(time.mktime(datetime.now().timetuple()))
        self.temperature = self.convert_temp(
            data["properties"]["temperature"]["values"][0]["value"]
        )
        self.min_temp = self.convert_temp(
            data["properties"]["minTemperature"]["values"][0]["value"]
        )
        self.max_temp = self.convert_temp(
            data["properties"]["maxTemperature"]["values"][0]["value"]
        )
        self.wind = wind
        self.aqi = 0
        self.feels_like = self.convert_temp(
            data["properties"]["apparentTemperature"]["values"][0]["value"]
        )
        self.dewpoint = data["properties"]["dewpoint"]["values"][0]["value"]
        self.cloud_cover = data["properties"]["skyCover"]["values"][0]["value"]
        conditions = []
        if self.cloud_cover == 0:
            conditions.append(WeatherCondition(800))
        elif self.cloud_cover < 25:
            conditions.append(WeatherCondition(801))
        elif self.cloud_cover < 50:
            conditions.append(WeatherCondition(802))
        elif self.cloud_cover < 85:
            conditions.append(WeatherCondition(803))
        else:
            conditions.append(WeatherCondition(804))
        if (
            0
            < data["properties"]["quantitativePrecipitation"]["values"][0]["value"]
            < 0.098
            and not metric
        ) or (
            0
            < data["properties"]["quantitativePrecipitation"]["values"][0]["value"]
            < 2.5
            and metric
        ):
            conditions.append(WeatherCondition(500))
        elif (
            data["properties"]["quantitativePrecipitation"]["values"][0]["value"] < 0.39
            and not metric
        ) or (
            data["properties"]["quantitativePrecipitation"]["values"][0]["value"] < 10
            and metric
        ):
            conditions.append(WeatherCondition(501))
        elif (
            data["properties"]["quantitativePrecipitation"]["values"][0]["value"] < 2
            and not metric
        ) or (
            data["properties"]["quantitativePrecipitation"]["values"][0]["value"] < 50
            and metric
        ):
            conditions.append(WeatherCondition(502))
        else:
            conditions.append(WeatherCondition(503))
        if len(data["properties"]["snowfallAmount"]["values"]) != 0:
            if data["properties"]["snowfallAmount"]["values"][0]["value"] != 0:
                conditions.append(WeatherCondition(601))
        self.conditions = conditions
        self.condition_ids = self.get_condition_ids()
        self.condition_sentence = self.get_conditions_sentence()

    def convert_temp(self, value):
        if self.metric:
            return value
        else:
            return value * 9 / 5 + 32

    def convert_speed(self, value):
        if self.metric:
            return value
        else:
            return value / 1.609

File: src/backend/nws/nws_current.rs
use crate::backend::nws::nws_json::NWSJSON;
use crate::backend::weather_condition::WeatherCondition;
use crate::backend::weather_data::{get_conditions_sentence, WeatherDataRS};
use crate::backend::WindData;
use crate::local::weather_file::WeatherFile;
use crate::now;

fn convert_temp(value: f64, metric: bool) -> f64 {
    if metric {
        value
    } else {
        value * 9.0 / 5.0 + 32.0
    }
}

fn convert_speed(value: f64, metric: bool) -> f64 {
    if metric { value } else { value * 0.62 }
}

fn get_conditions(
    data: NWSJSON,
    metric: bool,
    index: usize,
    cloud_cover: u8,
) -> crate::Result<Vec<WeatherCondition>> {
    let weather_file = WeatherFile::weather_codes()?;
    let mut conditions: Vec<WeatherCondition> = Vec::new();
    if cloud_cover == 0 {
        conditions.push(WeatherCondition::new(800, &weather_file.data)?);
    } else if cloud_cover < 25 {
        conditions.push(WeatherCondition::new(801, &weather_file.data)?);
    } else if cloud_cover < 50 {
        conditions.push(WeatherCondition::new(802, &weather_file.data)?);
    } else if cloud_cover < 85 {
        conditions.push(WeatherCondition::new(803, &weather_file.data)?);
    } else {
        conditions.push(WeatherCondition::new(804, &weather_file.data)?);
    }
    if data.properties.quantitative_precipitation.values[index].value != 0.0 {
        let rain = data.properties.quantitative_precipitation.values[index].value;
        let metric = metric;
        if (0.0 < rain && rain < 0.098 && !metric) || (0.0 < rain && rain < 2.5 && metric) {
            conditions.push(WeatherCondition::new(500, &weather_file.data)?);
        } else if (rain < 0.39 && !metric) || (rain < 10.0 && metric) {
            conditions.push(WeatherCondition::new(501, &weather_file.data)?);
        } else if (rain < 2.0 && !metric) || (rain < 50.0 && metric) {
            conditions.push(WeatherCondition::new(502, &weather_file.data)?);
        } else if rain != 0.0 {
            conditions.push(WeatherCondition::new(503, &weather_file.data)?);
        }
    }
    if data.properties.snowfall_amount.values[index].value != 0.0 {
        conditions.push(WeatherCondition::new(601, &weather_file.data)?);
    }
    Ok(conditions)
}

pub fn get_nws_current(data: NWSJSON, metric: bool) -> crate::Result<WeatherDataRS> {
    let cloud_cover = data.properties.sky_cover.values[0].value as u8;
    let conditions = get_conditions(data.clone(), metric, 0, cloud_cover)?;
    let d = WeatherDataRS {
        time: now() as i128,
        temperature: convert_temp(data.properties.temperature.values[0].value, metric)
            as f32,
        min_temp: convert_temp(
            data.properties.min_temperature.values[0].value,
            metric,
        ) as f32,
        max_temp: convert_temp(
            data.properties.max_temperature.values[0].value,
            metric,
        ) as f32,
        wind: WindData {
            speed: convert_speed(data.properties.wind_speed.values[0].value, metric),
            heading: data.properties.wind_direction.values[0].value as i16,
        },
        raw_data: String::new(),
        dewpoint: convert_temp(data.properties.dewpoint.values[0].value, metric) as f32,
        feels_like: convert_temp(
            data.properties.apparent_temperature.values[0].value,
            metric,
        ) as f32,
        aqi: 0,
        cloud_cover,
        conditions: vec![],
        condition_sentence: get_conditions_sentence(conditions),
    };
    Ok(d)
}
